psycopg3/transaction.py

```diff
@@ -0,0 +1,86 @@
+"""
+Transaction context managers returned by Connection.transaction()
+"""
+
+# Copyright (C) 2020 The Psycopg Team
+
+# TODO: Should this module be _underscored since it's implementation detail?
+from types import TracebackType
+from typing import Optional, Type, TYPE_CHECKING
+
+from .pq import TransactionStatus
+
+if TYPE_CHECKING:
+    from .connection import Connection
+
+
+class Transaction:
+    def __init__(self, conn: "Connection") -> None:
+        self._conn = conn
+        self._outer_transaction = False
+        self._savepoint_name: Optional[str] = None
+        self._original_autocommit: bool
+
+    @property
+    def connection(self) -> "Connection":
+        return self._conn
+
+    def __enter__(self) -> None:
+        with self._conn.lock:
+            if self._conn.pgconn.transaction_status == TransactionStatus.IDLE:
+                self._outer_transaction = True
+                self._savepoint_name = None
+
+                self._original_autocommit = self._conn.autocommit
+                if self._conn._autocommit:
+                    self._conn._autocommit = False
+
+                self._conn._start_query()
+                assert len(self._conn._savepoints) == 0
+                self._conn._savepoints.append(None)
+            else:
+                self._outer_transaction = False
+                self._savepoint_name = "tx_savepoint_{}".format(
+                    len([s for s in self._conn._savepoints if s is not None])
```

Comment:
This makes a O(1) operation into an O(n) operation. Why bothering counting the exactly non-null savepoints? You only need them unique (and we verified actually not even that is required... but let's say we want them unique).

How about `self._savepoint_name = f"tx_savepoint_{len(self._conn_savepoints)}"`?

Reply:
Yeah this was purely meant to avoid the curiosity of having your first savepoint named "tx_savepoint_2" in the case where an outer transaction is also started. This eye-bleed is elegantly solved through your suggestion of representing outer transaction with an empty list. +1